### PR TITLE
fix: update vocab size of language model's config on resize - mllama

### PR DESCRIPTION
**Pull request by Ssukriti**
# What does this PR do?

The PR updates the config of the mllama language model to update its vocab size. This is useful when vocab size of `MllamaForConditionalGeneration` class is updated when resizing embeddings using `model.resize_token_embeddings(embedding_size)` . **More details included in the issue below.**

The PR includes one potential fix. There are several ways to do this. Suggestions welcome!

Fixes # (issue)
https://github.com/huggingface/transformers/issues/36590

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Models:
- vision models: @amyeroberts, @qubvel

### CODE REVIEW

File: src/transformers/models/mllama/modeling_mllama.py [hunk 1]
```diff
@@ -1963,7 +1963,7 @@ def forward(
 
         loss = None
         if labels is not None:
-            loss = self.loss_function(logits, labels, self.vocab_size, **loss_kwargs)
+            loss = self.loss_function(logits, labels, self.config.vocab_size, **loss_kwargs)
```
Comment (zucchini-nlp) [inline src/transformers/models/mllama/modeling_mllama.py:1966]: let's make this `config.get_text_config().vocab_size`, that is the way it should be tbh. Then the below hacky reassignment will not be needed

Comment (Ssukriti) [reply src/transformers/models/mllama/modeling_mllama.py:1966]: I think the cleanest solution is to have a set_config_vocab_size function , so it can be called directly from resize https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L2219 , instead of `self.config.get_text_config().vocab_size = vocab_size`

and then in MllamaForConditionalGeneration we can just do

```
def set_config_vocab_size(self, ...):
        self.language_model.set_set_config_vocab_size(...)

```
this would make it similar to set_embeddings, but it will be a change in all model classes and the function implementation will be a one-liner only. 

Hence, I had just moved that one liner out to the forward() of `MllamaForConditionalGeneration` instead. Making that change might get easily missed for  multi-modal type wrapper classes though, which have a language model inside.

Comment (zucchini-nlp) [reply src/transformers/models/mllama/modeling_mllama.py:1966]: Ah that's a valid point. Yeah, in case of multimodal models it is impossible to infer where to properly set new vocab size. 

The approach with `self.set_vocab_size` looks good to me and we can use it in all types of multimodal models. The base case will be in `PreTrainedModel` class which defaults to simply setting `self.config.get_text_config().vocab_size = NEW_SIZE`

Will be happy to review changes. Also we'll need some tests for this in `tests/test_modeling_common.py`. LMK if you need any guidance

**Comment by Ssukriti** [comment]
@zucchini-nlp the Pr is ready for review changes have been made as discussed earlier (in resolved conversation above). 

1. all resize embeddings tests are passing, which indicates no change in current behavior 
2. added new unit tests for the function set_vocab_size 
3. mllama bug https://github.com/huggingface/transformers/issues/36590 is fixed. The llava class does not use vocab size in loss, hence the bug did not impact it. On quick look, I didnt find other applicable multimodel classes that could be impacted either. Hence, fix is only for mllama in this PR

**Review by zucchini-nlp — COMMENTED** [review]
Hey, sorry it took long to come back! 

After giving it more thought, I think the easiest solution is to stop passing `labels` to the language model and calculate loss from within `ConditionalGeneration` class. Something like https://github.com/huggingface/transformers/blob/a861db01e5df627cc4e6c414d23ef28a22d61efb/src/transformers/models/gemma3/modeling_gemma3.py#L1366-L1383, but with `self.loss_fn`

I realize this is not a perfect solution and doesn't scale, but we are planning big refactor on multimodal models, and the vocab size issue for labels will be resolved by then. Therefore I don't want us to add more abstraction which might be deleted later on

**Comment by Ssukriti** [comment]
@zucchini-nlp sure we can go with that approach instead . I created a new PR https://github.com/huggingface/transformers/pull/36840 as had to undo the changes here. 

If the solution looks ok, can continue with unit tests , thank you!

File: tests/models/mllama/test_modeling_mllama.py [hunk 2]
```diff
@@ -321,6 +321,22 @@ def test_inputs_embeds_matches_input_ids(self):
...
+        for model_class in self.all_model_classes:
+            model = model_class(config)
+            model.to(torch_device)
+            model.eval()
+
+            new_vocab_size = config.get_text_config().vocab_size + 10
+            model.set_vocab_size(new_vocab_size)
+            self.assertEqual(model.language_model.get_vocab_size(), new_vocab_size)
+            # model's get vocab returns language model's vocab size
+            self.assertEqual(model.language_model.get_vocab_size(), model.get_vocab_size())
```
[the remark this answers is not in the document]
Comment (Ssukriti) [reply tests/models/mllama/test_modeling_mllama.py]: Good call. I added the test in follow up PR https://github.com/huggingface/transformers/pull/36840

[the remark this answers is not in the document]
Comment (Ssukriti) [reply tests/models/mllama/test_modeling_mllama.py]: I also added you as a collaborator to my fork if needed. Invite is pending
